examples/synchronization/multi_function/ai_ao_sync.py

```diff
@@ -0,0 +1,94 @@
+from nidaqmx.constants import AcquisitionType
+
+
+def generate_sine_wave(
+    frequency: float,
+    amplitude: float,
+    sampling_rate: float,
+    number_of_samples: int,
+    phase_in: float = 0.0,
+) -> Tuple[numpy.typing.NDArray[numpy.double], float]:
+    """Generates a sine wave with a specified phase.
+
+    Args:
+        frequency (float): Specifies the frequency of the sine wave.
+        amplitude (float): Specifies the amplitude of the sine wave.
+        sampling_rate (float): Specifies the sampling rate of the sine wave.
+        number_of_samples (int): Specifies the number of samples to generate.
+        phase_in (Optional[float]): Specifies the phase of the sine wave in radians.
+
+    Returns:
+        Tuple[numpy.typing.NDArray[numpy.double], float]: Indicates a tuple
+        containing the generated data and the phase of the sine wave after generation.
+    """
+    duration_time = number_of_samples / sampling_rate
+    duration_radians = duration_time * 2 * np.pi
+    phase_out = (phase_in + duration_radians) % (2 * np.pi)
+    t = np.linspace(phase_in, phase_in + duration_radians, number_of_samples, endpoint=False)
+
+    return (amplitude * np.sin(frequency * t), phase_out)
+
+
+def main():
+    """Continuously acquires and generate data at the same time."""
+    total_read = 0
+    number_of_samples = 1000
+
+    with nidaqmx.Task() as ai_task, nidaqmx.Task() as ao_task:
+
+        def callback(task_handle, every_n_samples_event_type, number_of_samples, callback_data):
+            """Callback function for reading signals."""
+            nonlocal total_read
+            read = ai_task.read(number_of_samples_per_channel=number_of_samples)
+            total_read += len(read)
+            print(f"Acquired data: {len(read)} samples. Total {total_read}.", end="\r")
+
+            return 0
+
+        ai_task.ai_channels.add_ai_voltage_chan("Dev1/ai0")
+        ai_task.timing.cfg_samp_clk_timing(1000.0, sample_mode=AcquisitionType.CONTINUOUS)
+        ai_task.register_every_n_samples_acquired_into_buffer_event(1000, callback)
+        ao_task.ao_channels.add_ao_voltage_chan("Dev1/ao0")
+        ao_task.timing.cfg_samp_clk_timing(1000.0, sample_mode=AcquisitionType.CONTINUOUS)
+        ao_task.triggers.start_trigger.cfg_dig_edge_start_trig("ai/StartTrigger")
+
+        actual_sampling_rate = ao_task.timing.samp_clk_rate
+        print(f"Actual sampling rate: {actual_sampling_rate:g} S/s")
+
+        ao_data, _ = generate_sine_wave(
+            frequency=10.0,
+            amplitude=1.0,
+            sampling_rate=actual_sampling_rate,
+            number_of_samples=number_of_samples,
+        )
+
+        try:
+            ao_task.write(ao_data)
+            ao_task.start()
+            ai_task.start()
+
+            input("Acquiring samples continuously. Press Enter to stop.\n")
+
+        except nidaqmx.DaqError as e:
+            print(e)
+        finally:
+            ai_task.stop()
+            ao_task.stop()
```

Comment:
Is the try/except still necessary?

If start() raises an exception, the Python interpreter will print it, clear the task (which stops the task), and exit.

FYI: If read() raises an exception, the ctypes library (or nidaqmx-python's gRPC code) will print it and keep going.

Reply:
Try-except is not needed anymore after using the with statement.